altos-rust/port/cortex-m0/src/peripheral/usart/mod.rs

```diff
@@ -1,123 +1,305 @@
-// peripheral/serial/mod.rs
-// AltOSRust
-//
-// Created by Daniel Seitz on 11/30/16
+/*
+ * Copyright © 2017 AltOS-Rust Team
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+ */
+
 mod control;
 mod defs;
 mod baudr;
 mod tdr;
+mod rdr;
 mod isr;
+mod icr;
+
+/// This module is the highest level in the Usart heirarchy for implementing
+/// the serial driver.
+///
+/// Configuration for each of the two Usart registers, and each of the registers
+/// that are offset from Usartx, and the public functions used to initialize,
+/// configure, and manipulate the bits for each register is defined in this file.
+///
+/// The functions here are used as wrappers that pass the call down through
+/// each necessary level (one or more), until the actual register is reached
+/// and is able to set the bits for itself accordingly.
+///
+/// This module is also responsible for initial setup of the Usart register
+/// (Either Usart1 or Usart2)
 
 use super::{Control, Register};
 use volatile::Volatile;
 use self::control::UsartControl;
 use self::baudr::BRR;
 use self::tdr::TDR;
+use self::rdr::RDR;
 use self::isr::ISR;
+use self::icr::ICR;
 use self::defs::*;
 use peripheral::{rcc, gpio};
+use interrupt;
 
 pub use self::control::{WordLength, Mode, Parity, StopLength, HardwareFlowControl};
 pub use self::baudr::BaudRate;
 
-#[derive(Copy, Clone)]
+pub const USART2_TX_BUFFER_FULL_CHAN: usize = 43;
+pub const USART2_TC_CHAN: usize = 43 * 2;
+
+/// STM32F0 has two Usart registers available.
+#[derive(Copy, Clone, Debug)]
 pub enum UsartX {
     Usart1,
     Usart2,
```

Comment:
"Usart1 is connected to pins 9 (TX) and 10 (RX) of GPIO group A",
"Usart2 is connected to pins 2 (TX) and 3 (RX) of GPIO group A, and is connected to the debug port of the board"